src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1130,24 +1145,24 @@ ucs_status_t uct_ud_ep_flush(uct_ep_h ep_h, unsigned flags,
 
     uct_ud_enter(iface);
 
-    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
-        uct_ud_ep_purge(ep, UCS_ERR_CANCELED);
-        /* FIXME make flush(CANCEL) operation truly non-blocking and wait until
-         * all of the outstanding sends are completed. Without this, zero-copy
-         * sends which are still on the QP could be reported as completed which
-         * can lead to sending corrupt data, or local access error. */
-        status = UCS_OK;
-        goto out;
+    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL) &&
+        !(ep->flags & UCT_UD_EP_FLAG_FLUSH_CANCEL)) {
```

Comment:
should we really continue to uct_ud_ep_flush_nolock() if this is a 2nd cancel operation on same ep?

Reply:
why not? we should ensure completing provided completion in the current call after completing the previous one

Reply:
i see, because we have another uct_comp from user so need to add skb for it